scripts/docker-helper/cht-docker-compose.sh

```diff
@@ -0,0 +1,442 @@
+  if [ "$portIsOpen" = "0" ]; then
+    # todo - when thoere's no connectivity, this fails w/ DNS
+    # curl: (6) Could not resolve host: 127-0-0-1.my.local-ip.co
+
+    # todo - macos returns this error some times
+    # curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to 127-0-0-1.my.local-ip.co:8443
+    http_code=$(curl -k --silent --show-error --head "$chtUrl" --write-out '%{http_code}' | tail -n1)
+    if [ "$http_code" != "200" ]; then
+      echo "CHT is returning $http_code instead of 200."
+    fi
+  else
+    echo "Port $chtPort is not open on $chtIp"
+  fi
+}
+
+validate_env_file(){
+  envFile=$1
+  if [ ! -f "$envFile" ] || [[ ! "$(file $envFile)" == *"ASCII text"* ]]; then
+    echo "File not found or not a text file: $envFile"
+  else
+    # TODO- maybe grep for env vars we're expecting? Blindly including
+    # is a bit promiscuous
+    # shellcheck disable=SC1090
+    . "${envFile}"
+    if [ -z "$COMPOSE_PROJECT_NAME" ] || [ -z "$CHT_HTTP" ] || [ -z "$CHT_HTTPS" ]; then
+      echo "Missing env value in file: COMPOSE_PROJECT_NAME, CHT_HTTP or CHT_HTTPS"
+    elif [[ "$COMPOSE_PROJECT_NAME" =~ [A-Z] ]];then
+      echo "COMPOSE_PROJECT_NAME can not have upper case: $COMPOSE_PROJECT_NAME"
+    fi
+  fi
+}
+
+get_running_container_count(){
+  result=0
+  containers="$1"
+  IFS=' ' read -ra containersArray <<< "$containers"
+  for container in "${containersArray[@]}"
+  do
+    if [ "$( docker ps -f name="${container}" | wc -l )" -eq 2 ]; then
+        (( result++ ))
+    fi
+  done
+  echo "$result"
+}
+
+get_global_running_container_count(){
+  # shellcheck disable=SC1083
+  if [ "$( docker ps --format={{.Names}} | wc -l )" -gt 0 ]; then
+    docker ps --format={{.Names}} | wc -l
+  else
+    echo "0"
+  fi
+}
+
+volume_exists(){
+  project=$1
+  volume="${project}_medic-data"
+  if [ "$( docker volume inspect "${volume}" 2>&1  | wc -l )" -eq 2 ]; then
+    echo "0"
+  else
+    echo "1"
+  fi
+}
+
+get_images_count(){
+  images="$1"
+  result=0
+  IFS=' ' read -ra imagesArray <<< "$images"
+  for image in "${imagesArray[@]}"
+  do
+    if [ "$( docker image ls --format {{.Repository}}:{{.Tag}} | grep -c "${image}" )" -eq 1 ]; then
+        (( result++ ))
+    fi
+  done
+  echo "$result"
+}
+
+pull_images(){
+  images="$1"
+  IFS=' ' read -ra imagesArray <<< "$images"
+  for image in "${imagesArray[@]}"
+  do
+    docker image pull "${image}" >/dev/null 2>&1
+  done
+}
+
+get_docker_compose_yml_path(){
+  if [ -f docker-compose-developer.yml ]; then
+    echo "docker-compose-developer.yml"
+  elif [ -f ../../docker-compose-developer.yml ]; then
+    echo "../../docker-compose-developer.yml"
+  else
+    return 0
+  fi
+}
+
+docker_up_or_restart(){
+  # some times this function called too quickly after a docker change, so
+  # we sleep 3 secs here to let the docker container/volume stabilize
+  sleep 3
+
+  envFile=$1
+  composeFile=$2
+
+  # haproxy never starts on first "up" call, so you know, call it twice ;)
```

Comment:
Do we know why? This seems like something to fix.

Reply:
Indeed!  tl;dr is "we'll fix in arch v3". Part of the requirements for this ticket is to work under an unmodified arch v2 as that's what app devs are running in production today ([see related comment in parent ticket](https://github.com/medic/cht-core/issues/7061#issuecomment-825143864)). 

the too long is nicely covered by SRE when I asked about why Medic OS waits for HAproxy ( `depends_on: haproxy`) and not the other way around. My testing suggests that maybe switching this could avoid this "double started" you found in my code (thank you!!)  Their answer was really helpful in explaining why the helper script in this PR would be what's needed when we're stuck not fixing the underlying issue (again, the fix being upgrading to arch v3):

> API and sentinel route requests to couchdb through haproxy. So couchDB is supposed to run before haproxy and haproxy is supposed to run before api and sentinel. In arch v2, api, sentinel and couch are packaged in the medic-os container. Therefore parts of medic-os depend on haproxy and haproxy depends on other parts of medic-os. Meaning that config is not wrong and your change would also not be wrong. It all just depends on which processes start fast enough for this dependency loop to work correctly. This ambiguity isn’t there with arch v3. 
> 
> That lag and sort of race condition also becomes very noticeable when working with an instance where you’re trying to spawn with pre-existing data.